src/librustc_mir/dataflow/impls/borrows.rs

```diff
@@ -46,9 +47,118 @@ pub struct Borrows<'a, 'gcx: 'tcx, 'tcx: 'a> {
     root_scope: Option<region::Scope>,
 
     borrow_set: Rc<BorrowSet<'tcx>>,
+    borrows_out_of_scope_at_location: FxHashMap<Location, Vec<BorrowIndex>>,
 
     /// NLL region inference context with which NLL queries should be resolved
-    nonlexical_regioncx: Rc<RegionInferenceContext<'tcx>>,
+    _nonlexical_regioncx: Rc<RegionInferenceContext<'tcx>>,
+}
+
+fn precompute_borrows_out_of_scope<'a, 'tcx>(
+    mir: &'a Mir<'tcx>,
+    regioncx: &Rc<RegionInferenceContext<'tcx>>,
+    borrows_out_of_scope_at_location: &mut FxHashMap<Location, Vec<BorrowIndex>>,
+    borrow_index: BorrowIndex,
+    borrow_region: RegionVid,
+    location: Location,
+    visited_locations: &mut Vec<Location>
+) {
```

Comment:
Nit: I would definitely write this method with a `loop`. My usual formulation is something like this:

```rust
let mut stack = vec![location];
let mut visited = FxHashSet();
visited.insert(location);

while let Some(p) = stack.pop() {
    process(p);
    for each n in successors(p) {
        if visited.insert(n) {
            stack.push(n);
        }
    }
}
```

In this case, `process(p)` would be something like this:

```rust
if !regioncx.region_contains_point(borrow_region, location) {
    borrows_out_of_scope_at_location
      .entry(p)
      .or_insert_with(vec![])
      .push(borrow_index);
    continue; // do not visit successors of this point
}
```

and of course `successors(p)` is either a call to `Terminator::successors` or else `successor_within_block` — not sure if we have a more concise way to get *that* result right now...